markdown_parser.leg

```diff
@@ -223,6 +225,23 @@ BlockQuoteRaw =  a:StartList
                      $$->key = RAW;
                  }
 
+CommentBlock = a:CommentBlockRaw
+             {  $$ = mk_element(COMMENTBLOCK);
+                $$->children = a;
+             }
+
+CommentBlockRaw = a:StartList
+                ('<!--'
+				 (!'-->' CommentBlockLine { a = cons($$, a); } )*
+				'-->')+
+                 {   a = cons(mk_str("\033"), a); // add some non-whitespace character that will be stripped later. Required to keep the paragraph even for empty comments.
```

Comment:
Hier hab ich viel zu viel Zeit reingesteckt. Generell hab ich mich an `BlockQuote` gehalten, da es aber ein closing tag gibt, kann ich nicht einfach Zeilen parsen, die entweder in nem newline oder eof enden. 

Das Problem hier war, dass ich leere Kommentare (nur whitespace oder gar nichts) nicht importieren konnte. Die wurden erkannt, aber dann als komplett leer weggeschmissen. Bei leeren Block Quotes kommen da irgendwo her content, der das rausfiltern verhindert, am Ende aber auch nicht im Editor landet. Bei den comments hatte ich diese Zeichenketten nicht. Enthält das Element nur whitespace, fliegt es weg.
Ich versteh den Parsing Kram hier leider nur zu vielleicht 65%. Ich hab jetzt hier einen non-whitespace char (escape) eingefügt. Der ist nach dem Import im Editor nicht mehr da, aber das comment element überlebt den Import. Schön ists nicht, aber wollte erstmal die Reißleine ziehen, weil in das Detail zu viel Zeit reingeflossen ist. 

Vielleicht hast du da noch ne Idee?

Reply:
Ich hab das jetzt doch wieder rausnehmen können, weil ich endlich verstanden hab wie das bei Blockquotes funktioniert. Wir injecten das Zeichen vor dem Import, damit leere Quotes importiert werden können. Soweit ich das verstehe gibt das PEG nicht her, oder wir verstehens alle nicht. 
Ich injecte den special character jetzt auch für comments.

Reply:
Der Hintergrund ist, dass der Parser hier viele Whitespaces von sich aus rauswirft, weil er nur für die Konvertierung Markdown > HTML gedacht war. Wir müssen aber leere Elemente und Whitespaces erhalten, weil wir Roundtrip Safe sind. Damals wollten wir nicht den Parser komplett umschreiben, weil am Code von Multimarkdown dran bleiben wollten – daher passiert das über diese Injection. So generell könnte der PEG das schon.

Eines schönen Tages haben wir hoffentlich mal Zeit einen eigenen Markdown-Parser in Swift zu schreiben ;-). Oder Markdown ist bis dahin halbwegs bedeutungslos geworden :-D.